### ISSUE TITLE
no-descending-specificity trips up on pseudo-elements

### ISSUE THREAD

**Issue by Zirak** [stylelint/stylelint#4189]
> Clearly describe the bug

When choosing a selector to compare specificity to, `no-descending-specificity` discards any pseudo-classes like `last-child` before finding a selector to compare to. This seems sort of related to #2489, but from what I can tell that issue talks about more complex selectors where relationships may be tricky to define, while this has a more straightforward repro (see below).

If the pseudo-class was a regular class, I think we could agree that there'd be no reason to emit a warning, and intuitively seems like it should exhibit the same behaviour.

> Which rule, if any, is the bug related to?

`no-descending-specificity`

> What CSS is needed to reproduce the bug?

In `/tmp/blah.css`:

```css
.foo > span:last-child {
    color: #abc;
}

.bar > span {
    color: #fed;
}
```

> What stylelint configuration is needed to reproduce the bug?

In `/tmp/blah.stylelintrc.json`:

```json
{
    "rules": {
        "no-descending-specificity": true
    }
}
```

> Which version of stylelint are you using?

Reproduced in master:
```sh
$ ./bin/stylelint.js --version
10.1.0
```

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI from master, `./bin/stylelint.js --config /tmp/blah.stylelintrc.json /tmp/blah.css`

> What did you expect to happen?

No warnings to be flagged. The output should (imo) be equivalent to linting the following css:

```css
.foo > span.something-else {
    color: #abc;
}

.bar > span {
    color: #fed;
}
```

Whose output it empty.

> What actually happened (e.g. what warnings or errors did you get)?

```shell
/tmp/blah.css
 5:1  ✖  Expected selector ".bar > span" to come before selector ".foo > span:last-child"   no-descending-specificity
```

**Comment by nosilleg** [comment]
@zirak Thanks for the details in your issue, it makes things a lot easier to check.

The example you have provided works as intended. The specificity of those selectors is 0-2-1 and 0-1-1. Therefore the second is less specific and should come first.  The linting error is correct.

See https://www.codecaptain.io/tools/css-specificity-calculator for confirmation.

**Comment by Zirak** [comment]
@nosilleg My pleasure, thank you for taking a look !

You're right that the selector with the pseudo-class has a higher specificity, however so does the selector with a regular class. My intuition was that they'd be treated similarly. To better elaborate on what I mean:

```css
/* A: specificity 0-2-1 */
.foo span:last-child { ... }
/* B: specificity 0-2-1 */
.foo span.quz { ... }

/* C: specificity 0-1-1 */
.bar span { ... }
```

Currently, `no-descending-specificity` compares `A` and `C`, but does not compare `B` and `C`. It compares `A` and `C` because [it seems to strip all pseudo-classes](https://github.com/stylelint/stylelint/blob/4ee5fee06873248b7e23e4b54e2acce2868e7c72/lib/rules/no-descending-specificity/index.js#L97-L99), for better or for "worse".

**Comment by nosilleg** [comment]
Hi @Zirak 

Sorry for the delay in getting back to you.  Thankfully hudochenkov has added some labels anyway.

This does indeed look like a bug, with an linting error not being raised for:

```css
/* A: specificity 0-2-1 */
.foo span.quz { ... }

/* B: specificity 0-1-1 */
.bar span { ... }
```

Is this something that you think you can contribute to fixing?

**Comment by Andrioden** [comment]
Is this related to this false positive?

``` css
.a1 .b, .c:hover .b { margin-top: -1px }
.a2 .b, .c:hover .b { margin-left: -1px }
```
✖  Expected selector ".a2 .b" to come before selector ".c:hover .b"  no-descending-specificity

**Comment by nosilleg** [comment]
@Andrioden 

Firstly, CSS specificity is difficult to understand, which is why this rule exists.  The example you have provided is not a false positive, but is a correct error.

With you example, all rules apply to `.b`.

The [specificity calculations](https://www.codecaptain.io/tools/css-specificity-calculator) for your example are:

1. `.a1 .b`: 0,2,0
2. `.c:hover .b`: 0,3,0
3. `.a2 .b`: 0,2,0
4. `.c:hover .b`: 0,3,0

The `no-descending-specificity` rule says that a less specific rule should not come after a more specific rule.

So lines 2 and 3 are out of order, and that is what the error reported says.

In order to fulfil the requirements of this rule, your CSS should be:

```
.a1 .b { margin-top: -1px }
.a2 .b { margin-left: -1px }
.c:hover .b { margin-top: -1px ; margin-left: -1px }
```

This is more code, but it makes it more clear what the intention of your code is.  This is more clear because as far as CSS is concerned `{ margin-top: -1px ; margin-left: -1px }` can come about because of `.c:hover .b`, but also through `.a1 .a2 .b`, as well as other formations.

You as the author of the HTML may say that you would never make a `.a2` a descendant of `.a1`, but the CSS processor doesn't know that, and you have no way to communicate that to it.

As such the rule is doing it's job correctly in your example.

**Comment by Andrioden** [comment]
Thank you for that explantion, i did end up rewriting this exact css for other reasons, but i would do as you suggest if it still existed. I slightly agree that the css code is more clear with 3 lines, thou given my exact case (countering borders) you might argue grouping them is better, nevertheless, i agree. So consider my case closed. ^^ Not sure if its related to the OP thou and if the issue can be closed.
